File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        int length = 0;
        ListNode *curr = head;
        while(curr){
            curr = curr->next;
            length++;
        }
        int left = length-n;
        int count = 1;
        curr = head;
        while(count < left){
            curr = curr -> next;
            count++;
        }
        ListNode *temp = NULL;
        if(left == 0){
            temp = head -> next;
            head->next = NULL;
            delete head;
            head = temp;
        } else {
            temp = curr->next->next;
            curr->next->next = NULL;
            delete curr->next;
            curr->next = temp;
        }
        return head; 
    }
};

File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:

        dummyNode = ListNode()
        dummyNode.next = head

        slow = dummyNode
        fast = dummyNode
        while(n and fast):
            fast = fast.next
            n -= 1
        
        if fast is None:
            return head
        
        while(fast.next):
            slow = slow.next
            fast = fast.next
        
        temp = slow.next.next
        slow.next.next = None
        slow.next = temp

        return dummyNode.next
        